### PR TITLE
refactor: HG and AddWinsSet (put HG out of the CROs)

### PR DESCRIPTION
**Pull request by d-roak**

### CODE REVIEW

File: packages/object/src/hashgraph.ts [hunk 1]
```diff
@@ -17,105 +14,111 @@ export enum ActionType {
...
 		for (const dep of deps) {
 			if (!this.forwardEdges.has(dep)) {
-				this.forwardEdges.set(dep, new Set());
+				this.forwardEdges.set(dep, []);
 			}
-			this.forwardEdges.get(dep)?.add(hash);
-			this.frontier.delete(dep);
+			this.forwardEdges.get(dep)?.push(hash);
+
+			// delete from frontier frontier
```
Comment (guiltygyoza) [inline packages/object/src/hashgraph.ts:81]: redundant "frontier"

File: packages/node/src/index.ts [hunk 2]
```diff
@@ -90,31 +95,25 @@ export class TopologyNode {
 	}
 
 	updateObject(id: string, operations: { fn: string; args: string[] }[]) {
-		const object = TopologyObject.create({
+		// TODO: needs refactor for working with hash graph
```
Comment (guiltygyoza) [inline packages/node/src/index.ts:98]: what does this TODO entail?

Comment (guiltygyoza) [reply packages/node/src/index.ts:98]: trying to understand if this TODO belongs to this PR or not

**Review by sfroment — APPROVED** [review]

**Review by guiltygyoza — APPROVED** [review]
big refactor

File: packages/node/src/index.ts (hunk 2 above)
Comment (d-roak) [reply packages/node/src/index.ts:98]: no, next PR

File: packages/crdt/src/cros/AddWinsSet/index.ts [hunk 3]
```diff
@@ -1,72 +1,71 @@
...
 	add(value: T): void {
-		const op = new Operation(OperationType.Add, value);
-		this.state.set(value, (this.state.get(value) || 0) + 1);
+		this._add(value);
 	}
 
-	remove(value: T): void {
-		const op = new Operation(OperationType.Remove, value);
-		this.add(value);
+	private _remove(value: T): void {
```
[the remark this answers is not in the document]
Comment (d-roak) [reply packages/crdt/src/cros/AddWinsSet/index.ts:24]: cro's convergence is dependent on their HG (mix of resolveConflicts and mergeCallback), you don't care about the ops inside, they are just called locally

File: packages/crdt/src/cros/AddWinsSet/index.ts [hunk 4]
```diff
@@ -1,72 +1,71 @@
...
 
-	resolveConflicts(op1: Operation<T>, op2: Operation<T>): ActionType {
-		if (op1.type !== op2.type && op1.value === op2.value) {
-			return op1.type === OperationType.Add
-				? ActionType.DropRight
-				: ActionType.DropLeft;
-		}
-		return ActionType.Nop;
+	private _add(value: T): void {
+		if ((this.state.get(value) ?? 0) % 2 === 0) this.state.set(value, 1);
```
[the remark this answers is not in the document]
Comment (d-roak) [reply packages/crdt/src/cros/AddWinsSet/index.ts:17]: cro's convergence is dependent on their HG (mix of resolveConflicts and mergeCallback), you don't care about the ops inside, they are just called locally

**Pull request merged**